pkg/kubefedctl/federate/util.go

```diff
@@ -187,3 +192,35 @@ func getResourcesInNamespace(config *rest.Config, namespace, skipAPIResourceName
 
 	return resourcesInNamespace, nil
 }
+
+// decodeUnstructuredFromFile reads a list of yamls into a slice of unstructured objects
+func decodeUnstructuredFromFile(filename string) ([]*unstructured.Unstructured, error) {
+	f, err := os.Open(filename)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	var unstructuredList []*unstructured.Unstructured
+	reader := utilyaml.NewYAMLReader(bufio.NewReader(f))
+	for {
+		unstructuedObj := &unstructured.Unstructured{}
+		// Read one YAML document at a time, until io.EOF is returned
+		buf, err := reader.Read()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return nil, err
+		}
+		if len(buf) == 0 {
+			break
+		}
+		if err := yaml.Unmarshal(buf, unstructuedObj); err != nil {
+			return nil, err
+		}
```

Comment:
shall we use `NewYAMLToJSONDecoder` to simplify the logic here?

Reply:
Isn't reading a single yaml object just the degenerate case of reading multiple (just like splitting a string that doesn't contain a separator is effectively a no-op)?

Reply:
Theoretically what you mention is right. I did also expect the `decoder` (obtained by `NewYAMLToJSONDecoder`)  to ideally give me a single structure or a slice of structures or an unstructured list based on what it reads from the yaml file.
However the [decode](https://github.com/kubernetes-sigs/federation-v2/blob/master/vendor/k8s.io/apimachinery/pkg/util/yaml/decoder.go#L65) is not implemented this way. It decodes the whole yaml file content in one go which ends up decoding only the first object before the line separater (`---`). It does not return a list.
The utils provide an alternative way to obtain a [reader](https://github.com/kubernetes-sigs/federation-v2/blob/master/vendor/k8s.io/apimachinery/pkg/util/yaml/decoder.go#L262) which understands the separater (`---`)  and provides with individual yaml content on each read, which I have used here.